src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+            bar_str += "*";
+        }
+    } else {
+        for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+            int char_health = current_health - ( i * points_per_char );
+            if( char_health <= 0 ) {
+                bar_str += ".";
+            } else if( char_health == 1 ) {
+                bar_str += "\\";
+            } else {
+                bar_str += "|";
+            }
+        }
+    }
+
+    static const std::array<nc_color, 7> colors {{
+        c_green,
+        c_ltgreen,
+        c_yellow,
+        c_magenta,
+        c_ltred,
+        c_red,
+        c_ltgray
+    }};
+
+    bar_color = colors[(int) roundf( colors.size() * percentage )];
+}
+
+void health_bar::set_health_percentage( float percentage )
+{
+    bool overloaded = false;
+    if( percentage > 1 ) {
+        current_health = max_health;
+        overloaded = true;
+    } else if( percentage < 0 ) {
+        current_health = 0;
+    } else {
+        current_health = percentage * max_health;
+    }
+
+    refresh_bar( overloaded, percentage );
+}
+
+smiley_indicator::smiley_indicator( int x, int y, ui_anchor anchor) : ui_element(2, 1, x, y, anchor )
+{
+    set_state( neutral );
+}
+
+void smiley_indicator::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    mvwprintz( win, get_ay(), get_ax(), smiley_color, smiley_str.c_str() );
+}
+
+void smiley_indicator::set_state( smiley_state new_state )
+{
+    state = new_state;
+    switch( new_state ) {
+        case very_unhappy:
+            smiley_color = c_red;
+            smiley_str = "D:";
+            break;
+        case unhappy:
+            smiley_color = c_ltred;
+            smiley_str = ":(";
+            break;
+        case neutral:
+            smiley_color = c_white;
+            smiley_str = ":|";
+            break;
+        case happy:
+            smiley_color = c_ltgreen;
+            smiley_str = ":)";
+            break;
+        case very_happy:
+            smiley_color = c_green;
+            smiley_str = ":D";
+            break;
+    }
+}
+
+template<class T>
+tile_panel<T>::tile_panel( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor )
+                       : ui_element( size_x, size_y, x, y, anchor )
+{
+    num_tiles = size_x * size_y;
+    tiles = new T[num_tiles];
+}
+
+template<class T>
+void tile_panel<T>::set_rect( const ui_rect &new_rect )
+{
+    ui_element::set_rect( new_rect );
+    num_tiles = new_rect.size_x * new_rect.size_y;
+    delete[] tiles;
+    tiles = new T[num_tiles];
+}
+
+template<class T>
+tile_panel<T>::~tile_panel()
+{
+    delete[] tiles;
+}
+
+template<class T>
+void tile_panel<T>::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    for( unsigned int x = 0; x < get_rect().size_x; x++ ) {
+        for( unsigned int y = 0; y < get_rect().size_y; y++ ) {
+            tiles[y * get_rect().size_x + x].draw( win, x + get_ax(), y + get_ay() );
```

Comment:
It would be nice if `tile_panel` had a function/operator that returns a reference to the tile at (x,y). Only this function should be used to get the tile. In short: you want one place where the 2D-coordinate (x,y) is translated into the 1D-coordinate (the index into the vector).